File: codon/parser/visitors/typecheck/loops.cpp
// Copyright (C) 2022-2025 Exaloop Inc. <https://exaloop.io>

#include <string>
#include <tuple>

#include "codon/parser/ast.h"
#include "codon/parser/common.h"
#include "codon/parser/match.h"
#include "codon/parser/peg/peg.h"
#include "codon/parser/visitors/typecheck/typecheck.h"

using fmt::format;
using namespace codon::error;
namespace codon::ast {

using namespace types;
using namespace matcher;

/// Ensure that `break` is in a loop.
/// Transform if a loop break variable is available
/// (e.g., a break within loop-else block).
/// @example
///   `break` -> `no_break = False; break`

void TypecheckVisitor::visit(BreakStmt *stmt) {
  if (!ctx->getBase()->getLoop())
    E(Error::EXPECTED_LOOP, stmt, "break");
  ctx->getBase()->getLoop()->flat = false;
  if (!ctx->getBase()->getLoop()->breakVar.empty()) {
    resultStmt =
        N<SuiteStmt>(transform(N<AssignStmt>(
                         N<IdExpr>(ctx->getBase()->getLoop()->breakVar),
                         N<BoolExpr>(false), nullptr, AssignStmt::UpdateMode::Update)),
                     N<BreakStmt>());
  } else {
    stmt->setDone();
    if (!ctx->staticLoops.back().empty()) {
      auto a = N<AssignStmt>(N<IdExpr>(ctx->staticLoops.back()), N<BoolExpr>(false));
      a->setUpdate();
      resultStmt = transform(N<SuiteStmt>(a, stmt));
    }
  }
}

/// Ensure that `continue` is in a loop
void TypecheckVisitor::visit(ContinueStmt *stmt) {
  if (!ctx->getBase()->getLoop())
    E(Error::EXPECTED_LOOP, stmt, "continue");
  ctx->getBase()->getLoop()->flat = false;

  stmt->setDone();
  if (!ctx->staticLoops.back().empty()) {
    resultStmt = N<BreakStmt>();
    resultStmt->setDone();
  }
}

/// Transform a while loop.
/// @example
///   `while cond: ...`           ->  `while cond: ...`
///   `while cond: ... else: ...` -> ```no_break = True
///                                     while cond:
///                                       ...
///                                     if no_break: ...```
void TypecheckVisitor::visit(WhileStmt *stmt) {
  // Check for while-else clause
  std::string breakVar;
  if (stmt->getElse() && stmt->getElse()->firstInBlock()) {
    // no_break = True
    breakVar = getTemporaryVar("no_break");
    prependStmts->push_back(
        transform(N<AssignStmt>(N<IdExpr>(breakVar), N<BoolExpr>(true))));
  }

  ctx->staticLoops.push_back(stmt->gotoVar.empty() ? "" : stmt->gotoVar);
  ctx->getBase()->loops.emplace_back(breakVar);

  auto oldExpectedType = getStdLibType("bool")->shared_from_this();
  std::swap(ctx->expectedType, oldExpectedType);
  stmt->cond = transform(stmt->getCond());
  std::swap(ctx->expectedType, oldExpectedType);
  wrapExpr(&stmt->cond, getStdLibType("bool"));

  ctx->blockLevel++;
  stmt->suite = SuiteStmt::wrap(transform(stmt->getSuite()));
  ctx->blockLevel--;
  ctx->staticLoops.pop_back();

  // Complete while-else clause
  if (stmt->getElse() && stmt->getElse()->firstInBlock()) {
    auto es = stmt->getElse();
    stmt->elseSuite = nullptr;
    resultStmt = transform(N<SuiteStmt>(stmt, N<IfStmt>(N<IdExpr>(breakVar), es)));
  }
  ctx->getBase()->loops.pop_back();

  if (stmt->getCond()->isDone() && stmt->getSuite()->isDone())
    stmt->setDone();
}

/// Typecheck for statements. Wrap the iterator expression with `__iter__` if needed.
/// See @c transformHeterogenousTupleFor for iterating heterogenous tuples.
void TypecheckVisitor::visit(ForStmt *stmt) {
  if (stmt->isAsync())
    E(Error::CUSTOM, stmt, "async not yet supported");

  stmt->decorator = transformForDecorator(stmt->getDecorator());

  std::string breakVar;
  // Needs in-advance transformation to prevent name clashes with the iterator variable
  stmt->getIter()->setAttribute(
      Attr::ExprNoSpecial); // do not expand special calls here,
                            // might be needed for statis loops!
  stmt->iter = transform(stmt->getIter());

  // Check for for-else clause
  Stmt *assign = nullptr;
  if (stmt->getElse() && stmt->getElse()->firstInBlock()) {
    breakVar = getTemporaryVar("no_break");
    assign = transform(N<AssignStmt>(N<IdExpr>(breakVar), N<BoolExpr>(true)));
  }

  // Extract the iterator type of the for
  auto iterType = extractClassType(stmt->getIter());
  if (!iterType)
    return; // wait until the iterator is known

  auto [delay, staticLoop] = transformStaticForLoop(stmt);
  if (delay)
    return;
  if (staticLoop) {
    resultStmt = staticLoop;
    return;
  }

  // Replace for (i, j) in ... { ... } with for tmp in ...: { i, j = tmp ; ... }
  if (!cast<IdExpr>(stmt->getVar())) {
    auto var = N<IdExpr>(ctx->cache->getTemporaryVar("for"));
    auto ns = unpackAssignment(stmt->getVar(), var);
    stmt->suite = N<SuiteStmt>(ns, stmt->getSuite());
    stmt->var = var;
  }

  // Case: iterating a non-generator. Wrap with `__iter__`
  if (iterType->name != "Generator" && !stmt->isWrapped()) {
    stmt->iter = transform(N<CallExpr>(N<DotExpr>(stmt->getIter(), "__iter__")));
    iterType = extractClassType(stmt->getIter());
    stmt->wrapped = true;
  }

  ctx->getBase()->loops.emplace_back(breakVar);
  auto var = cast<IdExpr>(stmt->getVar());
  seqassert(var, "corrupt for variable: {}", *(stmt->getVar()));

  if (!var->hasAttribute(Attr::ExprDominated) &&
      !var->hasAttribute(Attr::ExprDominatedUsed)) {
    ctx->addVar(
        getUnmangledName(var->getValue()), ctx->generateCanonicalName(var->getValue()),
        stmt->getVar()->getType() ? stmt->getVar()->getType()->shared_from_this()
                                  : instantiateUnbound(),
        getTime());
  } else if (var->hasAttribute(Attr::ExprDominatedUsed)) {
    var->eraseAttribute(Attr::ExprDominatedUsed);
    var->setAttribute(Attr::ExprDominated);
    stmt->suite = N<SuiteStmt>(
        N<AssignStmt>(N<IdExpr>(format("{}{}", var->getValue(), VAR_USED_SUFFIX)),
                      N<BoolExpr>(true), nullptr, AssignStmt::UpdateMode::Update),
        stmt->getSuite());
  }
  stmt->var = transform(stmt->getVar());

  // Unify iterator variable and the iterator type
  if (iterType && iterType->name != "Generator")
    E(Error::EXPECTED_GENERATOR, stmt->getIter());
  if (iterType)
    unify(stmt->getVar()->getType(), extractClassGeneric(iterType));

  ctx->staticLoops.emplace_back();
  ctx->blockLevel++;
  stmt->suite = SuiteStmt::wrap(transform(stmt->getSuite()));
  ctx->blockLevel--;
  ctx->staticLoops.pop_back();

  if (ctx->getBase()->getLoop()->flat)
    stmt->flat = true;

  // Complete for-else clause
  if (stmt->getElse() && stmt->getElse()->firstInBlock()) {
    auto es = stmt->getElse();
    stmt->elseSuite = nullptr;
    resultStmt =
        transform(N<SuiteStmt>(assign, stmt, N<IfStmt>(N<IdExpr>(breakVar), es)));
    stmt->elseSuite = nullptr;
  }

  ctx->getBase()->loops.pop_back();

  if (stmt->getIter()->isDone() && stmt->getSuite()->isDone())
    stmt->setDone();
}

/// Transform and check for OpenMP decorator.
/// @example
///   `@par(num_threads=2, openmp="schedule(static)")` ->
///   `for_par(num_threads=2, schedule="static")`
Expr *TypecheckVisitor::transformForDecorator(Expr *decorator) {
  if (!decorator)
    return nullptr;
  Expr *callee = decorator;
  if (auto c = cast<CallExpr>(callee))
    callee = c->getExpr();
  auto ci = cast<IdExpr>(transform(callee));
  if (!ci || !startswith(ci->getValue(), "std.openmp.for_par.0")) {
    E(Error::LOOP_DECORATOR, decorator);
  }

  std::vector<CallArg> args;
  std::string openmp;
  std::vector<CallArg> omp;
  if (auto c = cast<CallExpr>(decorator))
    for (auto &a : *c) {
      if (a.getName() == "openmp" ||
          (a.getName().empty() && openmp.empty() && cast<StringExpr>(a.getExpr()))) {
        auto ompOrErr =
            parseOpenMP(ctx->cache, cast<StringExpr>(a.getExpr())->getValue(),
                        a.value->getSrcInfo());
        if (!ompOrErr)
          throw exc::ParserException(ompOrErr.takeError());
        omp = *ompOrErr;
      } else {
        args.emplace_back(a.getName(), transform(a.getExpr()));
      }
    }
  for (auto &a : omp)
    args.emplace_back(a.getName(), transform(a.getExpr()));
  return transform(N<CallExpr>(transform(N<IdExpr>("for_par")), args));
}

/// Handle static for constructs.
/// @example
///   `for i in statictuple(1, x): <suite>` ->
///   ```loop = True
///      while loop:
///        while loop:
///          i: Literal[int] = 1; <suite>; break
///        while loop:
///          i = x; <suite>; break
///        loop = False   # also set to False on break
/// If a loop is flat, while wrappers are removed.
/// A separate suite is generated for each static iteration.
std::pair<bool, Stmt *> TypecheckVisitor::transformStaticForLoop(ForStmt *stmt) {
  auto loopVar = getTemporaryVar("loop");
  auto suite = clean_clone(stmt->getSuite());
  auto [ok, delay, preamble, items] = transformStaticLoopCall(
      stmt->getVar(), &suite, stmt->getIter(), [&](Stmt *assigns) {
        Stmt *ret = nullptr;
        if (!stmt->flat) {
          auto brk = N<BreakStmt>();
          brk->setDone(); // Avoid transforming this one to skip extra checks
          // var [: Static] := expr; suite...
          auto loop = N<WhileStmt>(N<IdExpr>(loopVar),
                                   N<SuiteStmt>(assigns, clone(suite), brk));
          loop->gotoVar = loopVar;
          ret = loop;
        } else {
          ret = N<SuiteStmt>(assigns, clone(stmt->getSuite()));
        }
        return ret;
      });
  if (!ok)
    return {false, nullptr};
  if (delay)
    return {true, nullptr};

  // Close the loop
  auto block = N<SuiteStmt>();
  block->addStmt(preamble);
  for (auto &i : items)
    block->addStmt(cast<Stmt>(i));
  Stmt *loop = nullptr;
  if (!stmt->flat) {
    ctx->blockLevel++;
    auto a = N<AssignStmt>(N<IdExpr>(loopVar), N<BoolExpr>(false));
    a->setUpdate();
    block->addStmt(a);
    loop = transform(N<SuiteStmt>(N<AssignStmt>(N<IdExpr>(loopVar), N<BoolExpr>(true)),
                                  N<WhileStmt>(N<IdExpr>(loopVar), block)));
    ctx->blockLevel--;
  } else {
    loop = transform(block);
  }
  return {false, loop};
}

std::tuple<bool, bool, Stmt *, std::vector<ASTNode *>>
TypecheckVisitor::transformStaticLoopCall(Expr *varExpr, SuiteStmt **varSuite,
                                          Expr *iter,
                                          const std::function<ASTNode *(Stmt *)> &wrap,
                                          bool allowNonHeterogenous) {
  if (!iter->getClassType())
    return {true, true, nullptr, {}};

  std::vector<std::string> vars{};
  if (auto ei = cast<IdExpr>(varExpr)) {
    vars.push_back(ei->getValue());
  } else {
    Items<Expr *> *list = nullptr;
    if (auto el = cast<ListExpr>(varExpr))
      list = el;
    else if (auto et = cast<TupleExpr>(varExpr))
      list = et;
    if (list) {
      for (const auto &it : *list)
        if (auto ei = cast<IdExpr>(it)) {
          vars.push_back(ei->getValue());
        } else {
          return {false, false, nullptr, {}};
        }
    } else {
      return {false, false, nullptr, {}};
    }
  }

  Stmt *preamble = nullptr;
  iter = getHeadExpr(iter);
  auto fn =
      cast<CallExpr>(iter) ? cast<IdExpr>(cast<CallExpr>(iter)->getExpr()) : nullptr;
  std::vector<Stmt *> block;
  if (fn && startswith(fn->getValue(), "std.internal.static.tuple.0")) {
    block = populateStaticTupleLoop(iter, vars);
  } else if (fn && startswith(fn->getValue(), "std.internal.static.range.0:1")) {
    block = populateSimpleStaticRangeLoop(iter, vars);
  } else if (fn && startswith(fn->getValue(), "std.internal.static.range.0")) {
    block = populateStaticRangeLoop(iter, vars);
  } else if (fn &&
             startswith(fn->getValue(), "std.internal.static.function.0.overloads.0")) {
    block = populateStaticFnOverloadsLoop(iter, vars);
  } else if (fn && startswith(fn->getValue(), "std.internal.static.enumerate.0")) {
    block = populateStaticEnumerateLoop(iter, vars);
  } else if (fn && startswith(fn->getValue(), "std.internal.static.vars.0")) {
    block = populateStaticVarsLoop(iter, vars);
  } else if (fn && startswith(fn->getValue(), "std.internal.static.vars_types.0")) {
    block = populateStaticVarTypesLoop(iter, vars);
  } else {
    bool maybeHeterogenous = iter->getType()->is(TYPE_TUPLE);
    if (maybeHeterogenous) {
      if (!iter->getType()->canRealize())
        return {true, true, nullptr, {}}; // wait until the tuple is fully realizable
      if (!iter->getClassType()->getHeterogenousTuple() && !allowNonHeterogenous)
        return {false, false, nullptr, {}};
      block = populateStaticHeterogenousTupleLoop(iter, vars);
      preamble = block.back();
      block.pop_back();
    } else {
      return {false, false, nullptr, {}};
    }
  }
  std::vector<ASTNode *> wrapBlock;
  wrapBlock.reserve(block.size());
  for (auto b : block) {
    wrapBlock.push_back(wrap(b));
  }
  return {true, false, preamble, wrapBlock};
}

} // namespace codon::ast

File: meister/converted/passes/typecheck/loops.py
# Copyright (C) 2022-2026 Exaloop Inc. <https://exaloop.io>

from __future__ import annotations

from typing import TYPE_CHECKING

from ....bridge import Callable, List, cast
from ... import ast, cache
from . import assign, infer, special, utils
from .ctx import Base, TypecheckError

if TYPE_CHECKING:
    from . import TypeVisitor


def _parse_open_mp(code: str, info: ast.Node.SrcInfo) -> List[ast.CallExpr.Arg]:
    """Parse the OpenMP clauses accepted by codon/parser/peg/openmp.peg.
    Codex-generated."""

    # TODO: use proper parser?
    import re

    source = code.strip()
    source = re.sub(r"^omp\b", "", source, count=1).lstrip()
    source = re.sub(r"^parallel\b", "", source, count=1).lstrip()
    arguments = []
    clause = re.compile(
        r"(?:schedule\s*\(\s*(static|dynamic|guided|auto|runtime)"
        r"(?:\s*,\s*([1-9][0-9]*))?\s*\)"
        r"|num_threads\s*\(\s*([1-9][0-9]*)\s*\)"
        r"|ordered\b|collapse\s*\(\s*([1-9][0-9]*)\s*\)|gpu\b)"
    )
    while source:
        match = clause.match(source)
        if not match:
            raise TypecheckError(info, "openmp: invalid syntax")
        text = match.group(0)
        if text.startswith("schedule"):
            arguments.append(
                ast.CallExpr.Arg(name="schedule", value=ast.StringExpr(match.group(1), info=info))
            )
            if match.group(2):
                arguments.append(
                    ast.CallExpr.Arg(
                        name="chunk_size", value=ast.IntExpr(int(match.group(2)), info=info)
                    )
                )
        elif text.startswith("num_threads"):
            arguments.append(
                ast.CallExpr.Arg(
                    name="num_threads", value=ast.IntExpr(int(match.group(3)), info=info)
                )
            )
        elif text.startswith("ordered"):
            arguments.append(ast.CallExpr.Arg(name="ordered", value=ast.BoolExpr(True, info=info)))
        elif text.startswith("collapse"):
            arguments.append(
                ast.CallExpr.Arg(name="collapse", value=ast.IntExpr(int(match.group(4)), info=info))
            )
        else:
            arguments.append(ast.CallExpr.Arg(name="gpu", value=ast.BoolExpr(True, info=info)))
        source = source[match.end() :].lstrip()
    return arguments


def typecheck_break(self: TypeVisitor, node: ast.BreakStmt) -> ast.Node:
    """
    Ensure that `break` is in a loop.
    Transform if a loop break variable is available
    (e.g., a break within loop-else block).
    @example
    `break` -> `no_break = False; break`
    """

    loop = self.ctx.get_base().get_loop()
    if not loop:
        raise TypecheckError(node, "'break' outside loop")
    loop.flat = False
    if loop.break_var:
        assignment = ast.AssignStmt(
            ast.IdExpr(loop.break_var),
            rhs=ast.BoolExpr(False),
            update=ast.AssignStmt.Mode.Update,
        )
        assignment = self.visit(assignment)
        return ast.SuiteStmt(assignment, ast.BreakStmt())

    node.done = True
    if self.ctx.static_loops[-1]:
        assignment = ast.AssignStmt(
            ast.IdExpr(self.ctx.static_loops[-1]),
            rhs=ast.BoolExpr(False),
            update=ast.AssignStmt.Mode.Update,
        )
        return self.visit(ast.SuiteStmt(assignment, node))
    return node


def typecheck_continue(self: TypeVisitor, node: ast.ContinueStmt) -> ast.Node:
    """Ensure that `continue` is in a loop"""

    loop = self.ctx.get_base().get_loop()
    if not loop:
        raise TypecheckError(node, "'continue' outside loop")
    loop.flat = False

    node.done = True
    if self.ctx.static_loops[-1]:
        return ast.BreakStmt(done=True)
    return node


def typecheck_while(self: TypeVisitor, node: ast.WhileStmt) -> ast.Node:
    """
    Transform a while loop.
    @example
    `while cond: ...`           ->  `while cond: ...`
    `while cond: ... else: ...` -> ```no_break = True
    while cond:
    ...
    if no_break: ...```
    """

    # Check for while-else clause
    break_var = ""
    if node.else_suite and node.else_suite.first_in_block():
        # no_break = True
        break_var = utils.get_temporary_var(self.ctx, "no_break")
        assignment = self.visit(ast.AssignStmt(ast.IdExpr(break_var), rhs=ast.BoolExpr(True)))
        self.ctx.prepend_stmts[-1].append(assignment)

    base = self.ctx.get_base()
    base.loops.append(Base.LoopData(break_var=break_var))
    try:
        node.cond.expected_type = utils.get_stdlib_type(self.ctx, ast.types.Stdlib.Bool)
        node.cond = self.visit(node.cond)
        _, node.cond = utils.wrap_expr(
            self, node.cond, utils.get_stdlib_type(self.ctx, ast.types.Stdlib.Bool)
        )
        with (
            self.ctx.substitute("static_loops", self.ctx.static_loops + [node.goto_var or ""]),
            self.ctx.substitute("block_level", self.ctx.block_level + 1),
        ):
            node.suite = self.visit(node.suite)
            node.suite = ast.SuiteStmt.wrap(node.suite)
    finally:
        base.loops.pop()

    result: ast.Stmt = node
    # Complete while-else clause
    if node.else_suite and node.else_suite.first_in_block():
        suite, node.else_suite = node.else_suite, None
        result = self.visit(ast.SuiteStmt(node, ast.IfStmt(ast.IdExpr(break_var), if_suite=suite)))
    if node.cond.done and node.suite.done:
        node.done = True
    return result


def typecheck_for(self: TypeVisitor, node: ast.ForStmt) -> ast.Node:
    """
    Typecheck for statements. Wrap the iterator expression with `__iter__` if needed.
    See @c transformHeterogenousTupleFor for iterating heterogenous tuples.
    """

    if node.decorator:
        node.decorator = transform_for_decorator(self, node.decorator)
    match node.decorator:
        case ast.CallExpr(expr=ast.IdExpr(type=typ)) if (
            typ := typ.get_func()
        ) and typ.get_func_name() == ast.types.mangle("std.openmp", func="for_par"):
            static_bool = utils.extract_func_generic(typ, 3).get_bool_static()
            if static_bool and bool(static_bool.value):
                import_gpu = ast.ImportStmt(
                    ast.IdExpr("gpu"), args=[], as_=utils.get_temporary_var(self.ctx, "_")
                )
                self.ctx.prepend_stmts[-1].append(self.visit(import_gpu))

    break_var = ""
    # Needs in-advance transformation to prevent name clashes with the iterator variable
    # do not expand special calls here,
    node.iter.set(ast.Attr.ExprNoSpecial)
    # might be needed for static loops!
    node.iter = self.visit(node.iter)
    # Check for for-else clause
    assignment: ast.Stmt | None = None
    if node.else_suite and node.else_suite.first_in_block():
        break_var = utils.get_temporary_var(self.ctx, "no_break")
        assignment = self.visit(ast.AssignStmt(ast.IdExpr(break_var), rhs=ast.BoolExpr(True)))

    # Extract the iterator type of the for
    if not (iterator_type := node.iter.get_class_type()):
        return node

    delay, static_loop = transform_static_for_loop(self, node)
    if delay:
        return node  # wait until the iterator is known
    elif static_loop:
        return static_loop
    if not isinstance(node.var, ast.IdExpr):
        temp = ast.IdExpr(utils.get_temporary_var(self.ctx, "for"))
        unpacked = assign.unpack_assignment(self, node.var, temp)
        node.var = temp
        node.suite = ast.SuiteStmt(unpacked, node.suite)

    # Replace for (i, j) in ... { ... } with for tmp in ...: { i, j = tmp ; ... }
    is_generator = iterator_type.name == ("AsyncGenerator" if node.async_ else "Generator")
    if not is_generator and not node.wrapped:
        node.iter = self.visit(ast.CallExpr(ast.DotExpr(node.iter, member="__iter__")))
        iter_type = node.iter.get_class_type()
        node.wrapped = True
        if not iter_type:
            return node
        is_generator = iterator_type.name == ("AsyncGenerator" if node.async_ else "Generator")
    var = cast(ast.IdExpr, node.var)
    assert var, f"corrupt for variable: {node.var}"

    base = self.ctx.get_base()
    base.loops.append(Base.LoopData(break_var=break_var))
    try:
        if not var.has(ast.Attr.ExprDominated) and not var.has(ast.Attr.ExprDominatedUsed):
            var.type = var.type or utils.instantiate_unbound(self.ctx)
            self.ctx.add(
                utils.get_unmangled_name(self.ctx, var.value),
                self.ctx.generate_canonical_name(var.value),
                var.type,
                self.ctx.time,
            )
        elif var.has(ast.Attr.ExprDominatedUsed):
            var.erase(ast.Attr.ExprDominatedUsed)
            var.set(ast.Attr.ExprDominated)
            node.suite = ast.SuiteStmt(
                ast.AssignStmt(
                    ast.IdExpr(f"{var.value}{cache.VAR_USED_SUFFIX}"),
                    rhs=ast.BoolExpr(True),
                    update=ast.AssignStmt.Mode.Update,
                ),
                node.suite,
            )
        node.var = self.visit(var)

        # Case: iterating a non-generator. Wrap with `__iter__`
        if iterator_type and not is_generator:
            # Unify iterator var and the iterator type
            raise TypecheckError(node.iter, "expected iterable expression")
        if iterator_type:
            infer.unify(node.var.type, iterator_type[0])
        with (
            self.ctx.substitute("static_loops", self.ctx.static_loops + [""]),
            self.ctx.substitute("block_level", self.ctx.block_level + 1),
        ):
            node.suite = self.visit(node.suite)
            node.suite = ast.SuiteStmt.wrap(node.suite)
        if base.get_loop() and base.get_loop().flat:
            node.flat = True
        result = node
    finally:
        base.loops.pop()

    # Complete for-else clause
    if node.else_suite and node.else_suite.first_in_block():
        suite, node.else_suite = node.else_suite, None
        result = self.visit(
            ast.SuiteStmt(assignment, node, ast.IfStmt(ast.IdExpr(break_var), if_suite=suite))
        )
    if node.iter.done and node.suite.done:
        node.done = True
    return result


def transform_for_decorator(self: TypeVisitor, decorator: ast.Expr) -> ast.Expr:
    """
    Transform and check for OpenMP decorator.
    @example
    `@par(num_threads=2, openmp="schedule(static)")` ->
    `for_par(num_threads=2, schedule="static")`
    """

    callee = decorator
    if isinstance(callee, ast.CallExpr):
        callee = callee.expr
    callee = self.visit(callee)
    if not isinstance(callee, ast.IdExpr) or not callee.value.startswith(
        ast.types.mangle("std.openmp", func="for_par")
    ):
        raise TypecheckError(decorator, "invalid loop decorator")

    arguments = []
    omp_arguments = []
    if isinstance(decorator, ast.CallExpr):
        for arg in decorator.items:
            str_arg = arg.value if isinstance(arg.value, ast.StringExpr) else None
            if str_arg and (arg.name == "openmp" or not arg.name):
                omp_arguments = _parse_open_mp(str_arg.get_value(), str_arg.info)
            else:
                arguments.append(arg)
    arguments += omp_arguments
    result = self.visit(ast.CallExpr(ast.IdExpr("for_par"), items=arguments))
    return result


def transform_static_for_loop(self: TypeVisitor, stmt: ast.ForStmt):
    """
    Handle static for constructs.
    @example
    `for i in statictuple(1, x): <suite>` ->
    ```loop = True
    while loop:
    while loop:
    i: Literal[int] = 1; <suite>; break
    while loop:
    i = x; <suite>; break
    loop = False   # also set to False on break
    If a loop is flat, while wrappers are removed.
    A separate suite is generated for each static iteration.
    """

    loop_var = utils.get_temporary_var(self.ctx, "loop")
    suite = stmt.suite.clone(clean=True)

    def wrap(assignments: ast.Stmt) -> ast.Node:
        if not stmt.flat:
            break_stmt = ast.BreakStmt(done=True)  # set done to skip extra checks
            return ast.WhileStmt(
                ast.IdExpr(loop_var),
                suite=ast.SuiteStmt(assignments, suite.clone(), break_stmt),
                goto_var=loop_var,
            )
        else:
            return ast.SuiteStmt(assignments, stmt.suite.clone())

    ok, delay, preamble, items = transform_static_loop_call(self, stmt.var, stmt.iter, wrap)
    if not ok or delay:
        return ok, None
    block = ast.SuiteStmt(preamble, *items)
    if not stmt.flat:
        with self.ctx.substitute("block_level", self.ctx.block_level + 1):
            block.add(
                ast.AssignStmt(
                    ast.IdExpr(loop_var), rhs=ast.BoolExpr(False), update=ast.AssignStmt.Mode.Update
                )
            )
            # var [: Static] := expr; suite...
            loop_result = self.visit(
                ast.SuiteStmt(
                    ast.AssignStmt(ast.IdExpr(loop_var), rhs=ast.BoolExpr(True)),
                    ast.WhileStmt(ast.IdExpr(loop_var), suite=block),
                )
            )
    else:
        # Close the loop
        loop_result = self.visit(block)
    return False, loop_result


def populate_static_loop(
    self: TypeVisitor,
    var: ast.Expr | None,
    iterator: ast.Expr,
    final: ast.Expr,
) -> List[ast.Node]:
    results = []
    for idx in range(len(iterator.type.generics)):
        assignment = ast.AssignStmt(
            var.clone(), rhs=ast.IndexExpr(iterator.clone(), idx=ast.IntExpr(idx))
        )
        results.append(ast.StmtExpr([assignment], expr=final.clone))
    return results


def transform_static_loop_call(
    self: TypeVisitor,
    var: ast.Expr,
    iterator: ast.Expr,
    wrapper: Callable[[ast.Stmt], ast.Node],
    allow_non_heterogenous: bool = False,
):
    if not iterator.get_class_type():
        return True, True, None, []

    vars = []
    match var:
        case ast.IdExpr(value=name):
            vars.append(var.value)
        case ast.ListExpr(items=items) | ast.TupleExpr(items=items):
            if not items:
                return False, False, None, []
            for item in items:
                if isinstance(item, ast.IdExpr):
                    vars.append(item.value)
                else:
                    return False, False, None, []

    preamble: ast.Stmt | None = None
    iterator = utils.get_head_expr(iterator)
    function = (
        iterator.expr
        if isinstance(iterator, ast.CallExpr) and isinstance(iterator.expr, ast.IdExpr)
        else None
    )
    block: List[ast.Stmt] = []
    name = "" if not function else function.value
    if function and name.startswith(ast.types.mangle("std.internal.static", func="tuple")):
        block = special.populate_static_tuple_loop(self, iterator, vars)
    elif function and name.startswith(
        ast.types.mangle("std.internal.static", func="range", overload=1)
    ):
        block = special.populate_simple_static_range_loop(self, iterator, vars)
    elif function and name.startswith(ast.types.mangle("std.internal.static", func="range")):
        block = special.populate_static_range_loop(self, iterator, vars)
    elif function and name.startswith(
        ast.types.mangle("std.internal.static", cls="function", func="overloads")
    ):
        block = special.populate_static_fn_overloads_loop(self, iterator, vars)
    elif function and name.startswith(ast.types.mangle("std.internal.static", func="enumerate")):
        block = special.populate_static_enumerate_loop(self, iterator, vars)
    elif function and name.startswith(ast.types.mangle("std.internal.static", func="vars")):
        block = special.populate_static_vars_loop(self, iterator, vars)
    elif function and name.startswith(ast.types.mangle("std.internal.static", func="methods")):
        block = special.populate_static_methods_loop(self, iterator, vars)
    elif function and name.startswith(ast.types.mangle("std.internal.static", func="vars_types")):
        block = special.populate_static_var_types_loop(self, iterator, vars)
    elif isinstance(iterator.type, ast.types.Type) and iterator.type.is_type(
        ast.types.Stdlib.Tuple
    ):
        # Maybe heterogenous?
        if not iterator.type.can_realize():
            return True, True, None, []
        # wait until the tuple is fully realizable
        if not utils.is_heterogenous(self.ctx, iterator.type) and not allow_non_heterogenous:
            return False, False, None, []
        block = special.populate_static_heterogenous_tuple_loop(self, iterator, vars)
        preamble = block[-1]
        block.pop()
    else:
        return False, False, None, []
    return True, False, preamble, [wrapper(s) for s in block]
